docs/type-guards.md

```diff
@@ -0,0 +1,128 @@
+
+```lua
+type Tree = { value: number, left: Tree?, right: Tree? } | nil
+
+function isLeaf(x: Tree): x is { left: nil, right: nil }
+    return x ~= nil and x.left == nil and x.right == nil
+end
+
+local t: Tree = getTree()
+if isLeaf(t) then
+    print("Leaf:", t.value)
+    print(t.left, t.right)  -- Both known to be nil by the type solver
+end
+```
+
+## Design
+### Syntax
+The proposed syntax is for functions to optionally have a return type of `: x is T`, where `x` must be one of the arguments to the function, and `T` is the type the function narrows to. This amends the grammar to:
+
+```
+ReturnType ::= Type | TypePack | GenericTypePack | VariadicTypePack | TypePredicate
+TypePredicate ::= NAME 'is' Type
+```
+
+Type guards are allowed to be defined a members of tables, and the implicit `self` variable on methods is allowed to be used on self-call functions. The previous tree example could utilise this as:
+
+```lua
+local Tree = {}
+Tree.__index = Tree
+type Tree = setmetatable<{ value: number, left: Tree?, right: Tree? }, Tree>
+
+function Tree.new(value: number): Tree
+    return setmetatable({ value = value }, Tree)
+end
+function Tree:isLeaf(): self is { left: nil, right: nil }
+    return self.left == nil and self.right == nil
+end
+```
+
+### Semantics
+When type guards evaluate as true, the type at the call site should be intersected with the restriction present in the predicate. The above `isLeaf` function called as `if tree:isLeaf() then` results in a new type of `tree: typeof(tree) & { left: nil, right: nil }`, which will simplify to `{ @metatable Tree, { value: number, left: nil, right: nil } }`. (In reality this example currently ends up with some pretty nasty types, but it serves as illustration.)
+
+### Restrictions
+Type guards may only be used in the control flow statements `if`, `elseif` and `while`, the condition of an `if ... then ... else ...` expression, and assert statements. Retaining the value of a guard in a variable may lead to stale predicates no longer holding true.
+
+Multiple type guards may be combined using boolean operators. `and` performs intersection, `or` performs union, and `not` performs a compliment. This does not introduce new compliment logic to luau types, rather performing the same basic compliments that can be found in a statement such as `if (not (typeof(foo.x) == "number")) and foo.y == nil then`.
+
+Assigning the value of a type guard to a variable (`local foo = isCat(x)`) or its use in a more complex expression (`foo(isCat(x))`) is not disallowed, though the predicate returned by the type guard is demoted to a simple boolean value and no longer serves the narrow the type of the subject variable. It is suggested that lint rules may be used to warn about these cases.
+
+Type guard functions are not permitted to have multiple return values. `function foo(x): (x is number, string)` is disallowed, as is `function foo(x): (x is number, x is string)`.
```

Comment:
Looks a bit like an arbitrary choice.

Reply:
I wanted to allow predicates to be stored as normal variables, but consider what happens if you make a modification to `x` after the call to foo but before the use of your predicate booleans. Your predicates are now meaningless. We can't simply track things that might change, as the purpose of type guards is to also allow more complex logic than can be trivially identified by the type narrowing system. This is why the use of type functions was exclusively restricted to control flow statements, and as such any return values other than the first would be ignored. Disallowing multiple return values nips this footgun at the bud.

Reply:
I think that's actually okay. Type refinements store a proposition to refine a specific _version_ of a variable, so even if you update some variable and then apply an outdated proposition, it doesn't affect any variable whose version did not match. This gives us the effect of invalidating any outdated propositions.